tools/block-generator/generator/generate.go

```diff
@@ -369,90 +299,59 @@ func (g *generator) WriteBlock(output io.Writer, round uint64) error {
 		return nil
 	}
 	// round == nextRound case
+
 	err := g.startRound()
 	if err != nil {
 		return err
 	}
-	numTxnForBlock := g.txnForRound(g.round)
+	if g.round == 0 {
+		fmt.Printf("initializing generator with starting txnCounter from the genesis block: %d\n", g.txnCounter)
+	}
+	minTxnsForBlock := g.minTxnsForBlock(g.round)
 
 	var intra uint64 = 0
 	var cert rpcs.EncodedBlockCert
 	if g.round == 0 {
 		// we'll write genesis block / offset round for non-empty database
 		cert.Block, _, _ = g.ledger.BlockCert(basics.Round(round - g.roundOffset))
 	} else {
-		// generate a block
-		cert.Block.BlockHeader = bookkeeping.BlockHeader{
-			Round:          basics.Round(g.round),
-			Branch:         bookkeeping.BlockHash{},
-			Seed:           committee.Seed{},
-			TxnCommitments: bookkeeping.TxnCommitments{NativeSha512_256Commitment: crypto.Digest{}},
-			TimeStamp:      g.timestamp,
-			GenesisID:      g.genesisID,
-			GenesisHash:    g.genesisHash,
-			RewardsState: bookkeeping.RewardsState{
-				FeeSink:                   g.feeSink,
-				RewardsPool:               g.rewardsPool,
-				RewardsLevel:              0,
-				RewardsRate:               0,
-				RewardsResidue:            0,
-				RewardsRecalculationRound: 0,
-			},
-			UpgradeState: bookkeeping.UpgradeState{
-				CurrentProtocol: g.protocol,
-			},
-			UpgradeVote:        bookkeeping.UpgradeVote{},
-			StateProofTracking: nil,
-		}
+		g.setBlockHeader(&cert)
 
-		// Generate the transactions
-		transactions := []txn.SignedTxnInBlock{}
-		for intra < numTxnForBlock {
-			var signedTxns []txn.SignedTxn
-			var ads []txn.ApplyData
+		txGroupsAD := [][]txn.SignedTxnWithAD{}
+		for intra < minTxnsForBlock {
+			var txGroupAD []txn.SignedTxnWithAD
 			var err error
-			signedTxns, ads, intra, err = g.generateSignedTxns(g.round, intra)
+			txGroupAD, intra, err = g.generateTxGroup(g.round, intra)
 			if err != nil {
-				// return err
 				return fmt.Errorf("failed to generate transaction: %w", err)
 			}
-			if len(signedTxns) == 0 {
+			if len(txGroupAD) == 0 {
 				return fmt.Errorf("failed to generate transaction: no transactions given")
 			}
-			if len(signedTxns) != len(ads) {
-				return fmt.Errorf("failed to generate transaction: mismatched number of signed transactions (%d) and apply data (%d)", len(signedTxns), len(ads))
-			}
-			for i, stx := range signedTxns {
-				stib, err := cert.Block.BlockHeader.EncodeSignedTxn(stx, ads[i])
-				if err != nil {
-					return fmt.Errorf("failed to encode transaction: %w", err)
-				}
-				transactions = append(transactions, stib)
-			}
+			txGroupsAD = append(txGroupsAD, txGroupAD)
 		}
 
-		if intra < numTxnForBlock {
-			return fmt.Errorf("not enough transactions generated: %d > %d", numTxnForBlock, intra)
+		vBlock, ledgerTxnCount, err := g.evaluateBlock(cert.Block.BlockHeader, txGroupsAD, int(intra))
```

Comment:
Instead of calling `ledger.AddBlock()` by itself, we evaluate a block to generate a `*ledgercore.ValidatedBlock` and use it in `AddValidatedBlock()`

Reply:
Additionally, we read the new `ledgerTxnCount` from the evaluator's transaction counter.